lib/chef/data_collector.rb

```diff
@@ -467,21 +512,55 @@ def nested_resource?(new_resource)
         @current_resource_report && @current_resource_report.new_resource != new_resource
       end
 
+      def validate_and_return_uri(uri)
+        URI(uri)
+      rescue URI::InvalidURIError
+        return nil
+      end
+
+      def validate_and_create_file(file)
+        send_to_file_location(file, "")
+        return true
+      rescue Errno::ENOENT
+        return false
+      rescue Errno::EACCES
+        return false
+      end
+
       def validate_data_collector_server_url!
-        if data_collector_server_url.empty?
-          raise Chef::Exceptions::ConfigurationError,
-            "Chef::Config[:data_collector][:server_url] is empty. Please supply a valid URL."
-        end
+        unless !data_collector_server_url && data_collector_output_locations
+          uri = validate_and_return_uri(data_collector_server_url)
+          unless uri
+            raise Chef::Exceptions::ConfigurationError, "Chef::Config[:data_collector][:server_url] (#{data_collector_server_url}) is not a valid URI."
+          end
 
-        begin
-          uri = URI(data_collector_server_url)
-        rescue URI::InvalidURIError
-          raise Chef::Exceptions::ConfigurationError, "Chef::Config[:data_collector][:server_url] (#{data_collector_server_url}) is not a valid URI."
+          if uri.host.nil?
+            raise Chef::Exceptions::ConfigurationError,
+              "Chef::Config[:data_collector][:server_url] (#{data_collector_server_url}) is a URI with no host. Please supply a valid URL."
+          end
         end
+      end
 
-        if uri.host.nil?
+      def validate_data_collector_output_locations!
+        if data_collector_output_locations.empty?
           raise Chef::Exceptions::ConfigurationError,
-            "Chef::Config[:data_collector][:server_url] (#{data_collector_server_url}) is a URI with no host. Please supply a valid URL."
+                "Chef::Config[:data_collector][:output_loations] is empty. Please supply an hash of valid URLs and / or local file paths."
+        end
+
+        data_collector_output_locations.each do |type, locations|
+
+          method = if type == :urls
+                     method(:validate_and_return_uri)
+                   elsif type == :files
+                     method(:validate_and_create_file)
+                   end
```

Comment:
this is definitely not idiomatic ruby and looks like a C function pointer lookup table, and should probably get rewritten a bit cleaner, i'm going to need more coffee before i have a useful suggestion how to refactor it though.

Reply:
Feedback welcome on making this more idiomatic :)

Reply:
probably just:

```ruby
def handle_type(type, loc)
  type == :urls ? validate_and_return_uri(loc) : validate_and_create_file(loc)
end
```